### PR TITLE
[teammgrd]during warm-reboot teamd need to recover system-id from saved lacp-pdu

### PR DESCRIPTION
**Pull request by shine4chen**
Signed-off-by: shine.chen <shine.chen@nephosinc.com>

**What I did**
- When creating new portchannel, teammgrd gets the MAC from the saved LACP PDU and updates the port-channel MAC accordingly.
- Pls refer to [MCLAG-HLD](https://github.com/Azure/SONiC/pull/325) for details.

**Why I did it**
- In MC-LAG scenario, two peer devices form one end point of a LAG, these two devices must have the same MAC address since it’s used for LACP. During warm-reboot, this MAC must not be changed. 

**How I verified it**
- test it on nephos lab. With this approach we observe zero packet loss during mclag warm-reboot procedure.

**Details if related**

### CODE REVIEW

**Comment by lguohan** [comment]
vs test missing.

**Comment by shine4chen** [comment]
@lguohan  Teammgrd runs in teamd docker. Originally we want to add some vs test cases. But we find the existing vs image don't include teamd docker.  We can add vs test cases soon after teamd docker is added in vs image.

**Comment by adyeung** [comment]
BRCM team has completed the review, we have no further request or comment to add, please proceed to the next steps.

**Comment by pavel-shirshov** [comment]
@shine4chen 
How "correct" mac address is going to be set initially, before it is saved to a file?
Why mac address from config db is considered as incorrect?

**Comment by jianjundong** [comment]
> @shine4chen
> How "correct" mac address is going to be set initially, before it is saved to a file?
> Why mac address from config db is considered as incorrect?

@pavel-shirshov : Teamd saved the last PDU from partner per port in /var/warmboot/teamd/, the mac address of this device that saved in the PDU may be different to the mac address in config db. For example, in MCLAG scenario, the standby device will change the local mac address to the mac address of active device. If the standby device is warm rebooted, this mac address must be the same as that before warm-restart, if not, the portchannel will be broken and the data forwarding may be blocked.

File: cfgmgr/teammgr.cpp [hunk 1]
```diff
@@ -373,8 +375,33 @@ task_process_status TeamMgr::addLag(const string &alias, int min_links, bool fal
...
 
     stringstream conf;
+
+    const string dump_path = "/var/warmboot/teamd/";
+    MacAddress mac_boot = m_mac;
+    /*set portchannel mac same with mac before warmStart ,when warmStart there is a file written by teamd*/
+    ifstream aliasfile(dump_path + alias);
+    if (WarmStart::isWarmStart() && aliasfile.is_open())
+    {
+# define PARTNER_SYSTEM_ID_OFFSET 40
```
Comment (pavel-shirshov) [inline cfgmgr/teammgr.cpp]: I'd better to use `const auto` or `const int` here.

File: cfgmgr/teammgr.cpp [hunk 2]
```diff
@@ -373,8 +375,33 @@ task_process_status TeamMgr::addLag(const string &alias, int min_links, bool fal
     string res;
 
     stringstream conf;
+
+    const string dump_path = "/var/warmboot/teamd/";
+    MacAddress mac_boot = m_mac;
+    /*set portchannel mac same with mac before warmStart ,when warmStart there is a file written by teamd*/
```
Comment (pavel-shirshov) [inline cfgmgr/teammgr.cpp]: Please change it to:
`/* set portchannel mac same with mac before warmStart, when warmStart there is a file written by teamd */`

File: cfgmgr/teammgr.cpp [hunk 3]
```diff
@@ -373,8 +375,33 @@ task_process_status TeamMgr::addLag(const string &alias, int min_links, bool fal
...
+        while (getline(aliasfile, line))
+        {
+            ifstream memberfile(dump_path + line, ios::binary);
+            uint8_t mac_temp[ETHER_ADDR_LEN];
+
+            if (!memberfile.is_open())
+                continue;
+
+            memberfile.seekg(PARTNER_SYSTEM_ID_OFFSET, std::ios::beg);
+            memberfile.read((char *)mac_temp, ETHER_ADDR_LEN);
```
Comment (pavel-shirshov) [inline cfgmgr/teammgr.cpp]: instead of `(char *)` it's better to use `static_cast<char*>`

File: cfgmgr/teammgr.cpp (hunk 2 above)
Comment (shine4chen) [reply cfgmgr/teammgr.cpp]: thanks,done

File: cfgmgr/teammgr.cpp (hunk 1 above)
Comment (shine4chen) [reply cfgmgr/teammgr.cpp]: Thanks, done

File: cfgmgr/teammgr.cpp (hunk 3 above)
Comment (shine4chen) [reply cfgmgr/teammgr.cpp]: Thanks, done

**Comment by pavel-shirshov** [comment]
retest all please

**Comment by pavel-shirshov** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by pavel-shirshov** [comment]
@shine4chen Can you please check why checks are failing?

**Comment by shine4chen** [comment]
@pavel-shirshov I analysised 32 failed test cases and found two root cause.
1. cmd `ip link add pensive_ishi type veth peer name eth1` executes failed which result in the failure of `DockerVirtualSwitch`.
2. in this vs test environment [frr-vrf-patch](https://github.com/Azure/sonic-buildimage/pull/3763) is NOT imported which result in the failure of some vrf related test cases. I know frr-vrf-patch is merged to master. But for some environment issue the build failed, so vs test still gets the obsolete build artifact.

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by pavel-shirshov** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Pull request merged**